src/main.rs

```diff
@@ -0,0 +1,112 @@
+    while Path::new(&lock).exists() {
+        thread::sleep(time::Duration::from_secs(1));
+        if allocated == 0 {
+            let curr_free_ram_perc = get_free_ram_perc();
+            if curr_free_ram_perc < SWAPFC_FREE_PERC {
+                create_swapfile(&mut allocated);
+            }
+            continue;
+        }
+        let curr_free_swap_perc = get_free_swap_perc();
+        if curr_free_swap_perc < SWAPFC_FREE_PERC && allocated < SWAPFC_MAX_COUNT {
+            create_swapfile(&mut allocated);
+        }
+        if allocated <= 2 || allocated <= SWAPFC_MIN_COUNT {
+            continue;
+        }
+        if curr_free_swap_perc < SWAPFC_REMOVE_FREE_PERC {
+            destroy_swapfile(&mut allocated).expect("Unable to remove swap file");
+        }
+    }
+}
+
+fn swapfc_init(lock: &str) -> io::Result<()> {
+    fs::create_dir_all(SWAPFC_PATH).expect("Unable to create swapfc_path");
+//    fs::create_dir("/run/systemd/swap/swapfc").expect("Unable to create swapfc_run dir");
+    fs::File::create(lock).expect("Unable to create swapfc lock");
+    Ok(())
+}
+
+fn get_free_ram_perc() -> u8 {
+    let s = System::new_with_specifics(RefreshKind::new().with_memory());
+    let total = s.get_total_memory();
+    let free = s.get_free_memory();
+    ((free * 100) / total).try_into().unwrap()
+}
+
+fn get_free_swap_perc() -> u8 {
+    let s = System::new_with_specifics(RefreshKind::new().with_memory());
+    let total = s.get_total_swap();
+    let free = s.get_free_swap();
+    ((free * 100) / total).try_into().unwrap()
+}
+
+fn create_swapfile(allocated: &mut u8) -> () {
+//   if check_ENOSPC(swapfc_path)
+    sd_notify::notify(true, &[sd_notify::NotifyState::Status(String::from("Allocating swap file..."))]).expect("Unable to notify systemd");
+    *allocated += 1;
+    prepare_swapfile(*allocated).expect("Unable to prepare swap file");
+    Command::new("/usr/lib/systemd/systemd-makefs").arg("swap").arg(Path::new(SWAPFC_PATH).join(allocated.to_string())).output().expect("Unable to mkswap");
+    Command::new("/usr/bin/swapon").arg(SWAPFC_PATH.to_owned()+"/"+&allocated.to_string()).output().expect("Unable to swapon");
+    sd_notify::notify(true, &[sd_notify::NotifyState::Status(String::from("Monitoring memory status..."))]).expect("Unable to notify systemd");
+}
+
+fn prepare_swapfile(file: u8) -> io::Result<()> {
+    let mut file: String = file.to_string();
+    file = SWAPFC_PATH.to_owned()+"/"+&file;
+    // Verbose jargon to open 256M of /dev/zero
+    let mut src = fs::File::open("/dev/zero").expect("Unable to open file");
+    let mut buffer = vec![0; SWAPFC_CHUNK_SIZE];
+    src.read_exact(&mut buffer).expect("Unable to read to file");
+    // create swap file
+    fs::OpenOptions::new()
+        .create(true)
+        .write(true)
+        .mode(0o600)
+        .open(&file)
+        .unwrap();
+    // write to swap file
+    fs::write(&file, &mut buffer).expect("Unable to write file");
```

Comment:
This seems kinda wasteful. Doesn't rust allocate the entire buffer (256M by default) on the heap here before writing it out to the file all at once? If there isn't some magic going on because of it being all zeros, that is.

Reply:
I'm very very new to rust (second day). Could you precise a bit more?

Reply:
I did not write any meaningful rust code myself yet, but I am pretty sure, that

```rust
let mut buffer = vec![0; SWAPFC_CHUNK_SIZE];
```

allocates a Vec on the heap with the capacity of SWAPFC_CHUNK_SIZE (268435456) and initializes it with all zeros. I don't know what data type the rust compiler infers here, but I don't think it's u8 either. After that

```rust
src.read_exact(&mut buffer).expect("Unable to read to file");
```

reads buffer sized amount of zeros into the buffer.

```rust
fs::write(&file, &mut buffer).expect("Unable to write file");
```

This writes the whole buffer to the file, all at once. This is fine for small vectors but I don't think this is very efficient for such large chunks of data.

You are basically requesting 256 MiB RAM from the OS at a time when memory is very sparse.

Take this with a grain of salt since my exposure to rust is limited.

Reply:
You were completely right that we don't need to read from `/dev/zero`!
Pushed changes reflecting this.

Reply:
The main issue is the buffer allocation. I would recommend only allocating a maximum of 4 MiB and writing that repeatedly to the target file until you are done, if you roughly want to stick to your method of zeroing out the file.

Reply:
@zenofile I'll try that

Reply:
Implemented 4MiB buffer (according to [this](https://unix.stackexchange.com/questions/9432/is-there-a-way-to-determine-the-optimal-value-for-the-bs-parameter-to-dd) larger buffers don't really improve performance [note that I don't know how/(too lazy) to test this myself]).

Reply:
I think the resulting file will now be aligned to `SWAPFC_BUF_SIZE` boundaries. If `SWAPFC_CHUNK_SIZE` isn't a multiple of `SWAPFC_BUF_SIZE`, the resulting file size may differ from what the user did specify, albeit not by much.

Reply:
If you don't want to deal with this, you could always use `sendfile(2)` in conjunction with `/dev/zero`. Surely there is a rust wrapper for that system call.

Reply:
> Implemented 4MiB buffer (according to [this](https://unix.stackexchange.com/questions/9432/is-there-a-way-to-determine-the-optimal-value-for-the-bs-parameter-to-dd) larger buffers don't really improve performance [note that I don't know how/(too lazy) to test this myself]).

@Jarel1337 
```console
Finished release [optimized] target(s) in 0.27s
Running `target/release/zerofill`

1024 84.738928158s
4096 30.872728028s
8192 26.808183414s
16384 25.871767096s
4194304 38.411474791s
```
This was deliberately done on an older box (i7-2600k, 5.7.4-200.fc32.x86_64 with mitigations=off) on tmpfs, so dropping the page cache shouldn't be necessary. I guess this would be different with enabled mitigations on older systems, wrecking the context switching performance.

Code used: https://gist.github.com/zenofile/87480e14b32a76161eca1541f9f530e3 - with the resulting file size
 being exact and not aligned to the buffer size.